### PR TITLE
Extend aggregate function `first_value` and `last_value` 

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

extend `first_value` and `last_value` to accept NULL.

for example
```SQL
insert into test (a,b) values (1,null), (2,3), (4, 5), (6,null);
-- same as before
select first_value(b) from test;
-- if first encountered value is NULL, return NULL
select first_value(b) respect nulls from test;
```

```
┌─first_value_ignore_nulls(b)─┐
│              3 │
└────────────────┘

┌─first_value_respect_nulls(b)─┐
│                 ᴺᵁᴸᴸ │
└──────────────────────┘
```

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@lgbo-ustc please merge with https://github.com/ClickHouse/ClickHouse/pull/41067

**Comment by lgbo-ustc** [comment]
> @lgbo-ustc please merge with #41067

It makes a different implementation, but hase the same features. Should we add the new functions anyway?

**Comment by vdimir** [comment]
I'd suggest adding new function `first_value_with_null` (or something like) instead of parameter `first_value(true)`

**Comment by alexey-milovidov** [comment]
> It makes a different implementation, but hase the same features. Should we add the new functions anyway?

I prefer to have RESPECT NULLS and IGNORE NULLS modifiers, because they are SQL standard.
These modifiers can be implemented with a function, as in https://github.com/ClickHouse/ClickHouse/pull/41067

**Comment by lgbo-ustc** [comment]
@alexey-milovidov RESPECT NULLS and IGNORE NULLS  have been added.

File: src/Parsers/ExpressionListParsers.cpp [hunk 1]
```diff
@@ -1046,6 +1061,24 @@ class FunctionLayer : public Layer
...
+    {
+        DEFAULT = 0,
+        RESPECT_NULLS = 1,
+        IGNORE_NULLS = 2,
+    };
+    static String transformFunctionNameForRepectNulls(const String & original_function_name, NullsAction nulls_action)
+    {
+        static std::unordered_map<String, std::vector<String>> renamed_functions_with_nulls = {
+            {"first_value", {"first_value_ignore_nulls", "first_value_respect_nulls", "first_value_ignore_nulls"}},
+            {"last_value", {"last_value_ignore_nulls", "last_value_respect_nulls", "last_value_ignore_nulls"}},
```
Comment (vdimir) [inline src/Parsers/ExpressionListParsers.cpp:1075]: Can we leave current function name with no modifier? To have the same column name as before `first_value(x)` in such case

```suggestion
            {"first_value", {"first_value", "first_value_respect_nulls", "first_value_ignore_nulls"}},
            {"last_value", {"last_value", "last_value_respect_nulls", "last_value_ignore_nulls"}},
```

File: src/AggregateFunctions/AggregateFunctionMinMaxAny.h [hunk 2]
```diff
@@ -766,19 +766,24 @@ static_assert(
 
 
 /// For any other value types.
+template <bool IS_NULLABLE = false, bool IS_NULL_GREATER = false>
```
Comment (vdimir) [inline src/AggregateFunctions/AggregateFunctionMinMaxAny.h:769]: Do we use `IS_NULL_GREATER` ? If i get right it's always `false`

File: src/Parsers/ExpressionListParsers.cpp (hunk 1 above)
Comment (lgbo-ustc) [reply src/Parsers/ExpressionListParsers.cpp:1075]: The function factor use the first element to look up the function, it conflicts with previous `first_value`

File: src/AggregateFunctions/AggregateFunctionMinMaxAny.h (hunk 2 above)
Comment (lgbo-ustc) [reply src/AggregateFunctions/AggregateFunctionMinMaxAny.h:769]: It's false at present.

Comment (vdimir) [reply src/AggregateFunctions/AggregateFunctionMinMaxAny.h]: Let's remove it then?

**Pull request merged**